### PR TITLE
Feat: Cross-platform Build

### LINKED ISSUE #18: Add containerized build support
## Description

We should move the compile/build process to a container to ensure that everyone can get consistent products. Local compile/build often introduces problems due to inconsistent environment, such as inconsistent golang versions.

## Describe the Proposal

1. Adding golang:1.17.x as a builder container.
2. Updating Dockerfile & Makefile.
3. ...

### LINKED ISSUE #21: Cross-platform build support
## Description

As it's been discussed at https://github.com/merico-dev/stream/pull/20#discussion_r757267479, we need to support cross-platform build, but it's not very urgent.

We could support linux-amd64 darwin-amd64 windowns-amd64 first, then support other arch like "arm i386, mips64le, ppc64le, s390x ...

see also https://github.com/merico-dev/stream/pull/20

### PR DESCRIPTION
**Pull request by IronCore864**
# Summary

## Description

This PR supports building dtm and its plugins in a cross-platform manner. 

Src platform: darwin/arm64 (M1 Mac, aarch64)

| Destination platforms | Command to Build          | Method to Build | Why                                                                                          |
|-----------------------|---------------------------|-----------------|----------------------------------------------------------------------------------------------|
| darwin/arm64          | `make build-darwin-arm64` | local           | faster                                                                                       |
| darwin/amd64          | `make build-darwin-amd64` | local           | faster                                                                                       |
| linux/amd64           | `make build-linux-amd64`  | container       | plugin cross-platform build in container is a must https://github.com/golang/go/issues/22462 |

Other changes:

- updated readme and makefile comments
- created a `build` folder according to our [project layout doc](https://github.com/merico-dev/stream/blob/main/docs/project_layout.md#build)
- `make build-release` that builds for all platforms, useful when releasing a new version for all platforms.
- 

## Related Issues

Resolves #18 
Resolves #21 

## Screenshot

Test:

Screenshot 2022-01-11 at 12 17 18

### CODE REVIEW

**Comment by IronCore864** [comment]
The docker image is in my personal docker hub; will create a new repo under Merico and move it there.

File: Makefile [hunk 1]
```diff
@@ -9,10 +12,44 @@ build: fmt vet ## Build dtm & plugins locally.
...
+	mkdir -p .devstream
+	CGO_ENABLED=1 GOOS=darwin GOARCH=amd64 go build -buildmode=plugin -trimpath -gcflags="all=-N -l" -o .devstream/githubactions-darwin-amd64_0.0.1.so ./cmd/githubactions/
+	CGO_ENABLED=1 GOOS=darwin GOARCH=amd64 go build -buildmode=plugin -trimpath -gcflags="all=-N -l" -o .devstream/argocd-darwin-amd64_0.0.1.so ./cmd/argocd/
+	CGO_ENABLED=1 GOOS=darwin GOARCH=amd64 go build -buildmode=plugin -trimpath -gcflags="all=-N -l" -o .devstream/argocdapp-darwin-amd64_0.0.1.so ./cmd/argocdapp/
+	CGO_ENABLED=0 GOOS=darwin GOARCH=amd64 go build -trimpath -gcflags="all=-N -l" -o dtm-darwin-amd64 ./cmd/devstream/
+
+build-linux-amd64: ## Cross-platform build for linux/amd64
+	echo "Building in ${BUILD_PATH}"
+	rm -rf ${BUILD_PATH} && mkdir ${BUILD_PATH}
+	# docker buildx build --platform linux/amd64,linux/arm64 --push -t ironcore864/dtm-builder:v0.0.1 -f build/package/Dockerfile .
```
Comment (IronCore864) [inline Makefile:45]: Maybe we don't need to push the image to Docker hub.

**Review by daniel-hutao — APPROVED** [review]
/lgtm

**Review by lfbdev — APPROVED** [review]
/lgtm

**Pull request merged**